_posts/2019-12-10-variational-gaussian-process/vgp.Rmd

```diff
@@ -0,0 +1,549 @@
+fine_aggregate:age                    3.4524     1.2165   2.838 0.004657 ** 
+---
+Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
+
+Residual standard error: 8.327 on 788 degrees of freedom
+Multiple R-squared:  0.7656,	Adjusted R-squared:  0.7549 
+F-statistic: 71.48 on 36 and 788 DF,  p-value: < 2.2e-16
+```
+
+We also store the predictions on the test set, for later comparison.
+
+```{r}
+linreg_preds1 <- fit1 %>% predict(test[, 1:8])
+linreg_preds2 <- fit2 %>% predict(test[, 1:8])
+
+compare <-
+  data.frame(
+    y_true = test$strength,
+    linreg_preds1 = linreg_preds1,
+    linreg_preds2 = linreg_preds2
+  )
+```
+
+With no further preprocessing required, the [tfdatasets](https://tensorflow.rstudio.com/guide/tfdatasets/introduction/) input pipeline ends up nice and short:
+
+```{r}
+create_dataset <- function(df, batch_size, shuffle = TRUE) {
+  
+  df <- as.matrix(df)
+  ds <-
+    tensor_slices_dataset(list(df[, 1:8], df[, 9, drop = FALSE]))
+  if (shuffle)
+    ds <- ds %>% dataset_shuffle(buffer_size = nrow(df))
+  ds %>%
+    dataset_batch(batch_size = batch_size)
+  
+}
+
+# just one possible choice for batch size ...
+batch_size <- 64
+train_ds <- create_dataset(train, batch_size = batch_size)
+test_ds <- create_dataset(test, batch_size = nrow(test), shuffle = FALSE)
+```
+
+And on to model creation. 
+
+## The model
+
+Model definition is short as well, although there are a few things to expand on. Don't execute this yet:
+
+```{r}
+model <- keras_model_sequential() %>%
+  layer_dense(units = 8,
+              input_shape = 8,
+              use_bias = FALSE) %>%
+  layer_variational_gaussian_process(
+    # number of inducing points
+    num_inducing_points = num_inducing_points,
+    # kernel to be used by the wrapped Gaussian Process distribution
+    kernel_provider = RBFKernelFn(),
+    # output shape 
+    event_shape = 1, 
+    # initial values for the inducing points
+    inducing_index_points_initializer = initializer_constant(as.matrix(sampled_points)),
+    unconstrained_observation_noise_variance_initializer =
+      initializer_constant(array(0.1))
+  )
+
+```
+
+
+Two arguments to `layer_variational_gaussian_process()` need some preparation before we can actually run this. First, as the documentation tells us, `kernel_provider` should be 
+
+> a layer instance equipped with an \@property, which yields a `PositiveSemidefiniteKernel` instance". 
+
+In other words, the VGP layer wraps another Keras layer that, _itself_, wraps or bundles together the TensorFlow `Variables` containing the kernel parameters.
+
+We can make use of `reticulate`s new `PyClass` constructor to fulfill the above requirements.
+Using `PyClass`, we can directly inherit from a Python object, adding and/or overriding methods or fields as we like - and yes, even create a Python [property](https://docs.python.org/3/library/functions.html?#property).
+
+```{r}
+bt <- import("builtins")
+RBFKernelFn <- reticulate::PyClass(
```

Comment:
This look pretty rough... are we waiting for the python API to maybe implement something nicer before we try to wrap it?

Reply:
We will have convenience wrappers for layers, callbacks etc. in Keras eventually (see https://github.com/rstudio/keras/pull/942/)
Here, it really is a special case because it's only needed because of that `property` thing (which is a leaky implementation detail in TFP proper)